### PR TITLE
close #9691 and #10913('spawn'ed function cannot have a 'typed' or 'untyped' parameter)

### LINKED ISSUE #9691: `spawn echo()` results in internal error
### Example
```nim
import threadpool
spawn echo()
```

### Current Output
```
Hint: used config file 'config/nim.cfg' [Conf]
Hint: system [Processing]
Hint: test [Processing]
Hint: threadpool [Processing]
Hint: cpuinfo [Processing]
Hint: strutils [Processing]
Hint: parseutils [Processing]
Hint: math [Processing]
Hint: bitops [Processing]
Hint: algorithm [Processing]
Hint: unicode [Processing]
Hint: os [Processing]
Hint: posix [Processing]
Hint: times [Processing]
Hint: options [Processing]
Hint: typetraits [Processing]
Hint: strformat [Processing]
Hint: macros [Processing]
Hint: linux [Processing]
Hint: cpuload [Processing]
Error: internal error: getTypeDescAux(tyStmt)
```

### Additional Information
Nim version:

```
Nim Compiler Version 0.19.9 [Linux: amd64]
Compiled at 2018-11-12
Copyright (c) 2006-2018 by Andreas Rumpf

git hash: d2f7ecb80e8672f404fa59c9eb2249ffe72f5645
active boot switches: -d:release
```

### LINKED ISSUE #10913: Bleeding edge parallel example yields error in Nim 0.19.9
Example of parallel at https://nim-lang.github.io/Nim/manual.html#pragmas-experimental-pragma errors out when built:
`Error: internal error: (filename: "ccgexprs.nim", line: 999, column: 17)
Traceback from system (most recent call last)`

### Example
```nim
import threadpool

proc useParallel*[T](unused: T) =
  # use a generic T here to show the problem.
  {.push experimental: "parallel".}
  parallel:
    for i in 0..4:
      echo "echo in parallel"
  sync()
  
  {.pop.}

useParallel(1)
```

This complained of no `spawn` within `parallel` block, so I added a `spawn` before `echo`
```nim
import threadpool

proc useParallel*[T](unused: T) =
  # use a generic T here to show the problem.
  {.push experimental: "parallel".}
  parallel:
    for i in 0..4:
      spawn echo "echo in parallel"
  sync()
  
  {.pop.}

useParallel(1)
```

That's when I got the following output.

### Current Output
```bash
nim c -r --threads:on --verbosity:2 ../../t.nim
/etc/nim/nim.cfg(47, 3) Hint: added path: '/home/j/.nimble/pkgs/darwin-0.1.0' [Path]
    nimblepath="$home/.nimble/pkgs/"
    ^
/etc/nim/nim.cfg(47, 3) Hint: added path: '/home/j/.nimble/pkgs/opengl-#head' [Path]
    nimblepath="$home/.nimble/pkgs/"
    ^
/etc/nim/nim.cfg(47, 3) Hint: added path: '/home/j/.nimble/pkgs/nimBMP-0.1.6' [Path]
    nimblepath="$home/.nimble/pkgs/"
    ^
/etc/nim/nim.cfg(47, 3) Hint: added path: '/home/j/.nimble/pkgs/nimsl-0.3.1' [Path]
    nimblepath="$home/.nimble/pkgs/"
    ^
/etc/nim/nim.cfg(47, 3) Hint: added path: '/home/j/.nimble/pkgs/compiler-0.19.9' [Path]
    nimblepath="$home/.nimble/pkgs/"
    ^
/etc/nim/nim.cfg(47, 3) Hint: added path: '/home/j/.nimble/pkgs/variant-0.2.8' [Path]
    nimblepath="$home/.nimble/pkgs/"
    ^
/etc/nim/nim.cfg(47, 3) Hint: added path: '/home/j/.nimble/pkgs/sdl2-2.0' [Path]
    nimblepath="$home/.nimble/pkgs/"
    ^
/etc/nim/nim.cfg(47, 3) Hint: added path: '/home/j/.nimble/pkgs/android-0.1.0' [Path]
    nimblepath="$home/.nimble/pkgs/"
    ^
/etc/nim/nim.cfg(47, 3) Hint: added path: '/home/j/.nimble/pkgs/cairo-1.0' [Path]
    nimblepath="$home/.nimble/pkgs/"
    ^
/etc/nim/nim.cfg(47, 3) Hint: added path: '/home/j/.nimble/pkgs/ttf-0.2.11' [Path]
    nimblepath="$home/.nimble/pkgs/"
    ^
/etc/nim/nim.cfg(47, 3) Hint: added path: '/home/j/.nimble/pkgs/x11-1.1' [Path]
    nimblepath="$home/.nimble/pkgs/"
    ^
/etc/nim/nim.cfg(47, 3) Hint: added path: '/home/j/.nimble/pkgs/regex-0.10.1' [Path]
    nimblepath="$home/.nimble/pkgs/"
    ^
/etc/nim/nim.cfg(47, 3) Hint: added path: '/home/j/.nimble/pkgs/gtk2-1.3' [Path]
    nimblepath="$home/.nimble/pkgs/"
    ^
/etc/nim/nim.cfg(47, 3) Hint: added path: '/home/j/.nimble/pkgs/unicodedb-0.6.0' [Path]
    nimblepath="$home/.nimble/pkgs/"
    ^
/etc/nim/nim.cfg(47, 3) Hint: added path: '/home/j/.nimble/pkgs/nigui-0.1.1' [Path]
    nimblepath="$home/.nimble/pkgs/"
    ^
/etc/nim/nim.cfg(47, 3) Hint: added path: '/home/j/.nimble/pkgs/plists-0.1.0' [Path]
    nimblepath="$home/.nimble/pkgs/"
    ^
/etc/nim/nim.cfg(47, 3) Hint: added path: '/home/j/.nimble/pkgs/libffi-1.0.1' [Path]
    nimblepath="$home/.nimble/pkgs/"
    ^
/etc/nim/nim.cfg(47, 3) Hint: added path: '/home/j/.nimble/pkgs/os_files-0.1.1' [Path]
    nimblepath="$home/.nimble/pkgs/"
    ^
/etc/nim/nim.cfg(47, 3) Hint: added path: '/home/j/.nimble/pkgs/jnim-0.4.2' [Path]
    nimblepath="$home/.nimble/pkgs/"
    ^
/etc/nim/nim.cfg(47, 3) Hint: added path: '/home/j/.nimble/pkgs/nimx-0.1' [Path]
    nimblepath="$home/.nimble/pkgs/"
    ^
/etc/nim/nim.cfg(47, 3) Hint: added path: '/home/j/.nimble/pkgs/rect_packer-0.1.0' [Path]
    nimblepath="$home/.nimble/pkgs/"
    ^
/etc/nim/nim.cfg(47, 3) Hint: added path: '/home/j/.nimble/pkgs/winim-2.5.1' [Path]
    nimblepath="$home/.nimble/pkgs/"
    ^
/etc/nim/nim.cfg(47, 3) Hint: added path: '/home/j/.nimble/pkgs/jester-#head' [Path]
    nimblepath="$home/.nimble/pkgs/"
    ^
/etc/nim/nim.cfg(47, 3) Hint: added path: '/home/j/.nimble/pkgs/jsbind-0.1' [Path]
    nimblepath="$home/.nimble/pkgs/"
    ^
/etc/nim/nim.cfg(47, 3) Hint: added path: '/home/j/.nimble/pkgs/asynctools-0.1.0' [Path]
    nimblepath="$home/.nimble/pkgs/"
    ^
/etc/nim/nim.cfg(47, 3) Hint: added path: '/home/j/.nimble/pkgs/nimpy-0.1.0' [Path]
    nimblepath="$home/.nimble/pkgs/"
    ^
/etc/nim/nim.cfg(47, 3) Hint: added path: '/home/j/.nimble/pkgs/c2nim-0.9.14' [Path]
    nimblepath="$home/.nimble/pkgs/"
    ^
/etc/nim/nim.cfg(47, 3) Hint: added path: '/home/j/.nimble/pkgs/closure_compiler-0.3.4' [Path]
    nimblepath="$home/.nimble/pkgs/"
    ^
/etc/nim/nim.cfg(47, 3) Hint: added path: '/home/j/.nimble/pkgs/httpbeast-0.2.1' [Path]
    nimblepath="$home/.nimble/pkgs/"
    ^
/etc/nim/nim.cfg(47, 3) Hint: added path: '/home/j/.nimble/pkgs/async_http_request-0.1.0' [Path]
    nimblepath="$home/.nimble/pkgs/"
    ^
/etc/nim/nim.cfg(47, 3) Hint: added path: '/home/j/.nimble/pkgs/nake-1.9.3' [Path]
    nimblepath="$home/.nimble/pkgs/"
    ^
/etc/nim/nim.cfg(47, 3) Hint: added path: '/home/j/.nimble/pkgs/kiwi-0.1.0' [Path]
    nimblepath="$home/.nimble/pkgs/"
    ^
/etc/nim/nim.cfg(47, 3) Hint: added path: '/home/j/.nimble/pkgs/unicodeplus-0.4.0' [Path]
    nimblepath="$home/.nimble/pkgs/"
    ^
/etc/nim/nim.cfg(47, 3) Hint: added path: '/home/j/.nimble/pkgs/' [Path]
    nimblepath="$home/.nimble/pkgs/"
    ^
/etc/nim/nim.cfg(49, 5) Hint: added path: '/opt/nimble/pkgs/' [Path]
      nimblepath="/opt/nimble/pkgs/"
      ^
Hint: used config file '/etc/nim/nim.cfg' [Conf]
Hint: system [Processing]
Hint: widestrs [Processing]
Hint: io [Processing]
Hint: t [Processing]
Hint: threadpool [Processing]
Hint: cpuinfo [Processing]
Hint: strutils [Processing]
Hint: parseutils [Processing]
Hint: math [Processing]
Hint: bitops [Processing]
Hint: macros [Processing]
Hint: algorithm [Processing]
Hint: unicode [Processing]
Hint: os [Processing]
Hint: pathnorm [Processing]
Hint: osseps [Processing]
Hint: posix [Processing]
Hint: times [Processing]
Hint: options [Processing]
Hint: typetraits [Processing]
Hint: linux [Processing]
Hint: cpuload [Processing]
Error: internal error: (filename: "ccgexprs.nim", line: 999, column: 17)
Traceback from system (most recent call last)
```

### Expected Output
```
echo in parallel
echo in parallel
echo in parallel
echo in parallel
echo in parallel
```

### Possible Solution
Unknown

### Notes
It might be nice for the example to also echo which number parallel process it is currently on, the output of which might look something like this:
```
echo in parallel: 4
echo in parallel: 1
echo in parallel: 3
echo in parallel: 0
echo in parallel: 2
```

### Additional Information

Occurs on Linux virtual box Antergos instance
```
OS: Antergos
Kernel: x86_64 Linux 5.0.0-arch1-1-ARCH
CPU: Intel Core I7-6700 @ 2x 3.408GHz
GPU: vboxdrmfb
RAM: 1729MiB / 3000MiB

$ nim -v
Nim Compiler Version 0.19.9 [Linux: amd64]
Compiled at 2019-03-11
Copyright (c) 2006-2019 by Andreas Rumpf

git hash: 35aab357e2ad6ebe1d3b5196e592bba8a10bed83
active boot switches: -d:release -d:nativeStackTrace
```

And Windows host:
```
Edition: Windows 8.1 Pro
Processor: Intel(R) Core(TM) i7-6700 CPU @ 3.4GHz
Installed RAM: 16.0 GB (15.9 usable)

PS X:\> nim -v
Nim Compiler Version 0.19.9 [Windows: amd64]
Compiled at 2019-03-05
Copyright (c) 2006-2019 by Andreas Rumpf

git hash: aed8766e84364360e0ed1e1e667ee735afd99e81
active boot switches: -d:release
```

### PR DESCRIPTION
**Pull request by ringabout**
close #9691 
close #10913

### CODE REVIEW

File: compiler/spawn.nim [hunk 1]
```diff
@@ -233,6 +238,10 @@ proc setupArgsForParallelism(g: ModuleGraph; n: PNode; objType: PType;
     let n = n[i]
     if i < formals.len and formals[i].typ.kind in {tyStatic, tyTypeDesc}:
       continue
+
+    if formals[i].typ.kind == tyVarargs and formals[i].typ[0].kind in {tyTyped, tyUntyped} or 
+            formals[i].typ.kind in {tyTyped, tyUntyped}:
```
Comment (ringabout) [inline compiler/spawn.nim]:
```suggestion
    if formals[i].typ.kind == tyVarargs and (formals[i].typ[0].kind in {tyTyped, tyUntyped} or 
            formals[i].typ.kind in {tyTyped, tyUntyped}):
```

**Comment by Araq** [comment]
Why does it not support "echo" though? The wrapper proc is a first class entity. Is it because of varargs? Which procs do use typed/untyped anyway.

**Comment by ringabout** [comment]
The parameters of echo is typed, i'm not sure how to allocate memory for them(as the attributes of scratchObj). and if it is untyped, that's more difficult.Only magic procs can use typed/untyped parameter. So only a few procs are affected.

File: compiler/spawn.nim [hunk 2]
```diff
@@ -205,6 +205,11 @@ proc setupArgsForConcurrency(g: ModuleGraph; n: PNode; objType: PType;
     if i < formals.len:
       if formals[i].typ.kind in {tyVar, tyLent}:
         localError(g.config, n[i].info, "'spawn'ed function cannot have a 'var' parameter")
+
+      if formals[i].typ.kind == tyVarargs and (formals[i].typ[0].kind in {tyTyped, tyUntyped} or
+            formals[i].typ.kind in {tyTyped, tyUntyped}):
```
Comment (Araq) [inline compiler/spawn.nim]:
```suggestion
      if (formals[i].typ.kind == tyVarargs and formals[i].typ[0].kind in {tyTyped, tyUntyped}) or
            formals[i].typ.kind in {tyTyped, tyUntyped}:
```

File: compiler/spawn.nim [hunk 3]
```diff
@@ -233,6 +238,10 @@ proc setupArgsForParallelism(g: ModuleGraph; n: PNode; objType: PType;
     let n = n[i]
     if i < formals.len and formals[i].typ.kind in {tyStatic, tyTypeDesc}:
       continue
+
+    if formals[i].typ.kind == tyVarargs and (formals[i].typ[0].kind in {tyTyped, tyUntyped} or 
+            formals[i].typ.kind in {tyTyped, tyUntyped}):
```
Comment (Araq) [inline compiler/spawn.nim]: Same. --> Factor out into a helper proc. Copy&paste is bad programming style.

File: compiler/spawn.nim (hunk 2 above)
Comment (ringabout) [reply compiler/spawn.nim]: my mistake, fixed

File: compiler/spawn.nim (hunk 3 above)
Comment (ringabout) [reply compiler/spawn.nim]: done

File: compiler/spawn.nim [hunk 4]
```diff
@@ -192,6 +192,11 @@ proc createCastExpr(argsParam: PSym; objType: PType; idgen: IdGenerator): PNode
   result.typ = newType(tyPtr, nextTypeId idgen, objType.owner)
   result.typ.rawAddSon(objType)
 
+template checkMagicProcs(g: ModuleGraph, n: PNode, formals: PNode) =
```
Comment (Araq) [inline compiler/spawn.nim]:
```suggestion
proc checkMagicProcs(g: ModuleGraph, n: PNode, formal: PNode) =
```

File: compiler/spawn.nim [hunk 5]
```diff
@@ -192,6 +192,11 @@ proc createCastExpr(argsParam: PSym; objType: PType; idgen: IdGenerator): PNode
...
+  if (formals[i].typ.kind == tyVarargs and formals[i].typ[0].kind in {tyTyped, tyUntyped}) or
```
Comment (Araq) [inline compiler/spawn.nim]:
```suggestion
  if (formal.typ.kind == tyVarargs and formal.typ[0].kind in {tyTyped, tyUntyped}) or
```

File: compiler/spawn.nim [hunk 6]
```diff
@@ -192,6 +192,11 @@ proc createCastExpr(argsParam: PSym; objType: PType; idgen: IdGenerator): PNode
...
+          formals[i].typ.kind in {tyTyped, tyUntyped}:
```
Comment (Araq) [inline compiler/spawn.nim]:
```suggestion
          formal.typ.kind in {tyTyped, tyUntyped}:
```

File: compiler/spawn.nim [hunk 7]
```diff
@@ -206,9 +211,7 @@ proc setupArgsForConcurrency(g: ModuleGraph; n: PNode; objType: PType;
       if formals[i].typ.kind in {tyVar, tyLent}:
         localError(g.config, n[i].info, "'spawn'ed function cannot have a 'var' parameter")
 
-      if formals[i].typ.kind == tyVarargs and (formals[i].typ[0].kind in {tyTyped, tyUntyped} or
-            formals[i].typ.kind in {tyTyped, tyUntyped}):
-        localError(g.config, n[i].info, "'spawn'ed function cannot have a 'typed' or 'untyped' parameter")
+      checkMagicProcs(g, n[i], formals)
```
Comment (Araq) [inline compiler/spawn.nim]:
```suggestion
      checkMagicProcs(g, n[i], formals[i])
```

File: compiler/spawn.nim [hunk 8]
```diff
@@ -239,9 +242,8 @@ proc setupArgsForParallelism(g: ModuleGraph; n: PNode; objType: PType;
     if i < formals.len and formals[i].typ.kind in {tyStatic, tyTypeDesc}:
       continue
 
-    if formals[i].typ.kind == tyVarargs and (formals[i].typ[0].kind in {tyTyped, tyUntyped} or 
-            formals[i].typ.kind in {tyTyped, tyUntyped}):
-      localError(g.config, n.info, "'spawn'ed function cannot have a 'typed' or 'untyped' parameter")
+    checkMagicProcs(g, n, formals)
```
Comment (Araq) [inline compiler/spawn.nim]:
```suggestion
    checkMagicProcs(g, n, formals[i])
```

**Pull request merged**